File: katas/MarsRover/src/Rover.ts
import { Direction } from "./Direction";
import Controller from "./Controller";
import Grid from "./Grid";

export default class Rover {
  private controller: Controller;
  private grid: Grid;
  private direction: Direction;

  constructor(grid: Grid) {
    this.grid = grid;
    this.controller = new Controller();
    this.direction = Direction.North;
  }

  execute(command: string): string {
    const commands = command.split("");

    for (let i = 0; i < commands.length; i++) {
      const c = commands[i];
      if (this.move(c)) {
        this.grid.move(this.direction);
      } else if (this.rotate(c)) {
        this.direction = this.controller.rotate(c, this.direction);
      }
    }

    return `${this.grid.getX()}:${this.grid.getY()}:${this.direction}`;
  }

  private move(c: string) {
    return c === "M";
  }

  private rotate(c: string) {
    return c === "L" || c === "R";
  }
}

File: python/MarsRover/src/Rover.py
from src.Direction import Direction
from src.Grid import Grid
from src.Position import Position


class Rover:
    def __init__(self, grid):
        self.grid: Grid = grid
        self.position: Position = Position(0, 0)

    def execute(self, input):
        direction: Direction = Direction.NORTH
        commands = [ch for ch in input]

        for c in commands:
            if (c == 'R'):
                direction = self.turnRight(direction)
            elif(c == 'L'):
                direction = self.turnLeft(direction)
            elif(c == 'M'):
                self.position = self.grid.move(direction, self.position)


        return f"{self.position.x}:{self.position.y}:{direction.value}"

    def turnRight(self, direction):
        switcher = {
            Direction.NORTH: Direction.EAST,
            Direction.EAST: Direction.SOUTH,
            Direction.SOUTH: Direction.WEST,
            Direction.WEST: Direction.NORTH
        }
        return switcher.get(direction, Direction.NORTH)

    def turnLeft(self, direction):
        switcher = {
            Direction.NORTH: Direction.WEST,
            Direction.WEST: Direction.SOUTH,
            Direction.SOUTH: Direction.EAST,
            Direction.EAST: Direction.NORTH
        }
        return switcher.get(direction, Direction.NORTH)
